Feature/testexecution

gunther-bachmann opened the pull request:
GitProvider is pretty printed, sorry that this ended up in this pull request

Commit:
configure ssh session factory

gunther-bachmann committed:
initial simple executor with tests

gunther-bachmann committed:
pretty print

jjelschen commented on org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/TestExecutorProvider.xtend:
```diff
@@ -0,0 +1,46 @@
+
+	def ProcessBuilder testExecutionBuilder(String testCase) {
+		if (!testCase.endsWith(TEST_CASE_FILE_SUFFIX)) {
+			logger.warn('''File '«testCase»' is no test case (does not end on «TEST_CASE_FILE_SUFFIX»)''')
```
isn't this actually an error?

jjelschen commented on org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/TestExecutorProvider.xtend:
```diff
@@ -0,0 +1,46 @@
+		val testFile = workspaceProvider.workspace.toPath.resolve(testCase)
+		if (!testFile.toFile.exists) {
+			logger.error('''File '«testCase»' does not exist''')
+			return null
```
smells bad… throw exception instead?

jjelschen commented on org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/TestExecutorResource.xtend:
```diff
@@ -0,0 +1,37 @@
+	@Path("execute") 
+	def Response executeTests(@QueryParam("resource") String resourcePath) {
+		val builder = executorProvider.testExecutionBuilder(resourcePath)
+		if (builder !== null) {
```
I think if you throw exceptions in `TestExecutorProvider.testExecutionBuilder`, you will not only be able to drop the null test, but can also rely on Jersey to take care of responding with a server error that should contain the exception message and associated stack trace.

jjelschen commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorTest.xtend:
```diff
@@ -0,0 +1,104 @@
+	WorkspaceProvider workspaceProviderMock
+
+	@InjectMocks
+	TestExecutorProvider testExecutorProvider // class under test
```
Why not call the field itself `unitUnderTest` or `testExecutorUnderTest`?

jjelschen commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorTest.xtend:
```diff
@@ -0,0 +1,104 @@
+
+	@Test
+	def void testNonExistingFile() {
+
```
// given
    val nonExistingFile = 'fantasy-file.tcl'
    …
?
(just for readability)

jjelschen commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorTest.xtend:
```diff
@@ -0,0 +1,104 @@
+	def void testWithRegularTestFile() {
+		// given 
+		val actualFile = 'src/test/java/org/example/actual-file.tcl'
+		temporaryFolders.newFolder(actualFile.split('/').takeWhile[!endsWith('.tcl')])
```
looks a bit over-sophisticated, but I can't come up with a better / shorter alternative :grinning:

gunther-bachmann committed:
cleaned up based on review

gunther-bachmann committed:
added logfile creation (piping stdout into file)

franzbecker commented on org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/TestExecutorProvider.xtend:
```diff
@@ -0,0 +1,69 @@
+
+	@Inject WorkspaceProvider workspaceProvider
+
+	def ProcessBuilder testExecutionBuilder(String testCase) {
```
I think this message is too big and does too much

franzbecker commented on org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/TestExecutorProvider.xtend:
```diff
@@ -0,0 +1,69 @@
+		.redirectError(Redirect.INHERIT)
+
+		processBuilder.environment.put(LOGFILE_ENV_KEY, logFile)
+		new File(workingDir + '/' + logFile).parentFile.mkdirs
```
it even creates directories..

franzbecker commented on org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/TestExecutorProvider.xtend:
```diff
@@ -0,0 +1,69 @@
+		return processBuilder
+	}
+
+	private def String logFileName(String testClass) '''«LOG_FOLDER»/testrun-«testClass»-«new SimpleDateFormat('yyyyMMddHHmmSSS').format(Date.newInstance)».log'''
```
This is pretty compact. I'd prefer a while method which also uses the new date API (LocalDate)

franzbecker commented on org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/TestExecutorResource.xtend:
```diff
@@ -0,0 +1,33 @@
+	@Path("execute")
+	def Response executeTests(@QueryParam("resource") String resourcePath) {
+		val builder = executorProvider.testExecutionBuilder(resourcePath)
+		logger.info('''Starting test for '«resourcePath»'.''')
```
We should use the proper logging pattern here

franzbecker commented on org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/TestExecutorResource.xtend:
```diff
@@ -0,0 +1,33 @@
+	@Inject TestExecutorProvider executorProvider
+
+	@POST
+	@Consumes(MediaType.APPLICATION_JSON)
```
This seems redundant (already specified on class level)

franzbecker commented on org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/TestExecutorResource.xtend:
```diff
@@ -0,0 +1,33 @@
+		val builder = executorProvider.testExecutionBuilder(resourcePath)
+		logger.info('''Starting test for '«resourcePath»'.''')
+		builder.start
+		return Response.created(UriBuilder.fromResource(DocumentResource).build(#[builder.environment.get(TestExecutorProvider.LOGFILE_ENV_KEY)], false)).build
```
That line is too hard to read

franzbecker commented on org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/TestExecutorResource.xtend:
```diff
@@ -0,0 +1,33 @@
+
+	@Inject TestExecutorProvider executorProvider
+
+	@POST
```
Why is it a post and not a get?

gunther-bachmann replied:
Since a testexecution is created. Idea is that a test execution itself is a resource that can be created (may be even deleted, etc.)

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorIntegrationTest.xtend:
```diff
@@ -0,0 +1,47 @@
+		]
+
+		// when
+		val request = createTestExecutionRequest(testFile).buildPost(Entity.entity('{ }', MediaType.APPLICATION_JSON_TYPE))
```
Why do we need to pass an empty json here?

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorIntegrationTest.xtend:
```diff
@@ -0,0 +1,47 @@
+		// then
+		assertThat(response.status).isEqualTo(Status.CREATED.statusCode)
+		val logFileURI = response.headers.getFirst(HttpHeaders.LOCATION) as String
+		val logAsRelativeFile = logFileURI.split('/').dropWhile[!equals('documents')].drop(1).join('/')
```
I'd feel more comfortable with a regex... something like `logFileURI.replaceFirst('.*/documents/, '')`

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorIntegrationTest.xtend:
```diff
@@ -0,0 +1,47 @@
+		val logAsRelativeFile = logFileURI.split('/').dropWhile[!equals('documents')].drop(1).join('/')
+		val logfile = workspaceRoot.root.toPath.resolve(userId + '/' + logAsRelativeFile).toFile
+		assertThat(logfile).exists
+		val executionResult = workspaceRoot.root.toPath.resolve(userId + '/test.ok.txt').toFile
```
`workspaceRoot.root.toPath.resolve` is kind of duplicate code here

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorTest.xtend:
```diff
@@ -0,0 +1,118 @@
+class TestExecutorTest {
+
+	@Rule
+	public var temporaryFolders = new TemporaryFolder
```
`val`

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorTest.xtend:
```diff
@@ -0,0 +1,118 @@
+	WorkspaceProvider workspaceProviderMock
+
+	@InjectMocks
+	TestExecutorProvider testExecutorProviderUnderTest // class under test
```
order - should be above the other mock

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorTest.xtend:
```diff
@@ -0,0 +1,118 @@
+	TestExecutorProvider testExecutorProviderUnderTest // class under test
+
+	@Mock
+	var Appender<ILoggingEvent> logAppender
```
`var` can be left out here

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorTest.xtend:
```diff
@@ -0,0 +1,118 @@
+		// when
+		try {
+			testExecutorProviderUnderTest.testExecutionBuilder('actual-file.tl')
+			Assert.fail
```
`Assert.fail('expected exception was not thrown')`

gunther-bachmann committed:
rework based on review

- cleanup of TestExecutorProvider
  - postpone (optional) folder creation to actual test execution
  - make log file name creation more explicit
  - use new date time api
- remove redundant/unnecessary Consumes/Produces from TestExectorResource
- clean up Tests
  - make TestExectorResource less obscure
  - give test names more expectation information
  - remove unnessecary json payload
  - reorder such that class under test comes first
  - added failure messages when expecting exceptions

franzbecker commented on org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/TestExecutorProvider.xtend:
```diff
@@ -0,0 +1,69 @@
+			throw new IllegalArgumentException(errorMsg)
+		}
+
+		val testClass = testCase.toTestClassName
```
All this code above could be extracted in a method called `getTestClass(String testCase)`

franzbecker commented on org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/TestExecutorProvider.xtend:
```diff
@@ -0,0 +1,69 @@
+		val testClass = testCase.toTestClassName
+		val logFile = testClass.createNewLogFileName
+		val workingDir = workspaceProvider.workspace.absoluteFile
+		val processBuilder = new ProcessBuilder() //
```
I'd prefer something like:
```
val processBuilder = new ProcessBuilder => [
  command(...)
  directory(...)
   ...
]

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorTest.xtend:
```diff
@@ -0,0 +1,118 @@
+	WorkspaceProvider workspaceProviderMock
+	@Mock
+	Appender<ILoggingEvent> logAppender
+	ArgumentCaptor<LoggingEvent> logCaptor
```
Can't we use `@Captor` here?

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorTest.xtend:
```diff
@@ -0,0 +1,118 @@
+
+	@Before
+	def void setupMocks() {
+		MockitoAnnotations.initMocks(this)
```
If we'd reuse an abstract test class we wouldn't need this line?!

gunther-bachmann replied:
no other test uses this, yet, so i'd prefer to not have an abstract test class (yet)?!

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorTest.xtend:
```diff
@@ -0,0 +1,118 @@
+class TestExecutorTest {
+
+	@Rule
+	public val temporaryFolders = new TemporaryFolder
```
should be singular

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorTest.xtend:
```diff
@@ -0,0 +1,118 @@
+
+		processBuilder => [
+			assertThat(command).haveAtLeastOne(
+				new Condition<String>([matches('^.*\\./gradlew .*--tests org\\.example\\.actual-file .*$')], 'gradle command line with test class'))
```
hard to read

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorTest.xtend:
```diff
@@ -0,0 +1,118 @@
+			assertThat(command).haveAtLeastOne(
+				new Condition<String>([matches('^.*\\./gradlew .*--tests org\\.example\\.actual-file .*$')], 'gradle command line with test class'))
+			assertThat(environment).hasEntrySatisfying(TestExecutorProvider.LOGFILE_ENV_KEY, new Condition<String>([
+				matches(TestExecutorProvider.LOG_FOLDER + '/testrun-org\\.example\\.actual-file-[0-9]{15}\\.log')
```
hard to read

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorIntegrationTest.xtend:
```diff
@@ -0,0 +1,48 @@
+		val logAsRelativeFile = logFileURI.replaceFirst('.*/documents/','')
+		val workspaceRootPath = workspaceRoot.root.toPath
+		val logfile = workspaceRootPath.resolve(userId + '/' + logAsRelativeFile).toFile
+		assertThat(logfile).exists
```
code above should be refactored into a method

gunther-bachmann committed:
reworked based on review #2

- Used Captor Annotation instead of manually setting
- renamed to singluar (temporaryFolder)
- moved log file name extraction from header in separate method
- moved some string matching boilerplate into separate methods

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorTest.xtend:
```diff
@@ -0,0 +1,129 @@
+
+	@Mock
+	WorkspaceProvider workspaceProviderMock
+	@Mock
```
If we have many annotated fields I personally prefer having them on the same line. Alternatively, we should be consistent with the blank lines between the fields.

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorTest.xtend:
```diff
@@ -0,0 +1,129 @@
+			Assert.fail('expected exception was NOT thrown!')
+		} catch (IllegalArgumentException exception) {
+			// then
+			assertThat(exception.message).contains(''''«existingNonTestFile»' is no test case'''.toString)
```
I think using `ExpectectException` would be a bit more readable and also allows us to verify the message.

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorTest.xtend:
```diff
@@ -0,0 +1,129 @@
+
+		processBuilder => [
+			assertThat(command).haveAtLeastOne(
+				stringMatching(regExContainingInOrder('./gradlew', 'test', '--tests org.example.actual-file'), 'gradle command line with test class'))
```
Could you assign the call to `regExContainingInOrder` to a variable first?

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorTest.xtend:
```diff
@@ -0,0 +1,129 @@
+				stringMatching(regExContainingInOrder('./gradlew', 'test', '--tests org.example.actual-file'), 'gradle command line with test class'))
+
+			assertThat(environment).hasEntrySatisfying(TestExecutorProvider.LOGFILE_ENV_KEY,
+				stringMatching('\\Q' + TestExecutorProvider.LOG_FOLDER + '/testrun-org.example.actual-file-\\E[0-9]{15}\\.log', 'log file named accordingly'))
```
same here I still think this line is too long

franzbecker commented on org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/TestExecutorTest.xtend:
```diff
@@ -0,0 +1,129 @@
+	@Test
+	def void testWithRegularTestFileCreatesWellFormedProcessBuilder() {
+		// given
+		val actualFile = 'src/test/java/org/example/actual-file.tcl'
```
Why not `ExampleTest.tcl` ?

gunther-bachmann committed:
rework based on review #3

- test extends Abstract Test (no mock initialization anymore)
- field annotations now inline
- exceptions now named expectedExceptions
- test file names carry more semantics now
- asserts shortened, regexes extracted into 'expected' vals
